File: Algorithms/DivideConquer/QuickHull.java

import java.awt.Point;
import java.util.HashSet;

public class QuickHull {

    private static int getLineDist(Point a, Point b, Point x) {
        return (x.y - a.y) * (b.x - a.x) - (b.y - a.y) * (x.x - a.x);
    }

    private static void findHull(Point[] points, HashSet<Point> hull, Point a, Point b, int side) {

        int maxIndex = -1, maxDist = 0;

        for (int i = 0; i < points.length; i++) {

            int dist = getLineDist(a, b, points[i]);

            int curDist = Math.abs(dist);
            int curSide = Integer.compare(dist, 0);

            if (curSide == side && curDist > maxDist) {

                maxIndex = i;
                maxDist = curDist;

            }

        }

        if (maxIndex == -1) {

            hull.add(a);
            hull.add(b);

        } else {

            findHull(points, hull, points[maxIndex], a, Integer.compare(0, getLineDist(points[maxIndex], a, b)));
            findHull(points, hull, points[maxIndex], b, Integer.compare(0, getLineDist(points[maxIndex], b, a)));

        }

    }

    public static HashSet<Point> getHull(Point[] points) {

        HashSet<Point> hull = new HashSet<>();

        int minX = 0, maxX = 0;

        for (int i = 0; i < points.length; i++) {

            if (points[i].x < points[minX].x) {
                minX = i;
            }

            if (points[i].x > points[maxX].x) {
                maxX = i;
            }

        }

        findHull(points, hull, points[minX], points[maxX], 1);
        findHull(points, hull, points[minX], points[maxX], -1);

        return hull;

    }

    public static void main(String[] args) {

        Point[] points = new Point[]{
            new Point(0, 3),
            new Point(1, 1),
            new Point(2, 2),
            new Point(4, 4),
            new Point(0, 0),
            new Point(1, 2),
            new Point(3, 1),
            new Point(3, 3)
        };

        for (Point p : getHull(points)) {
            System.out.println(p);
        }

    }

}

File: Algorithms/DivideConquer/QuickHull.js
"use strict";

let getLineDist = (a, b, x) => {
    return (x.y - a.y) * (b.x - a.x) - (b.y - a.y) * (x.x - a.x);
}

let findHull = (points, hull, a, b, side) => {

    let maxIndex = -1,
        maxDist = 0;

    for (let [index, point] of points.entries()) {

        let dist = getLineDist(a, b, point);

        let curDist = Math.abs(dist);
        let curSide = Math.sign(dist);

        if (curSide == side && curDist > maxDist) {

            maxIndex = index;
            maxDist = curDist;

        }

    }

    if (maxIndex == -1) {

        hull.add(a);
        hull.add(b);

    } else {

        findHull(points, hull, points[maxIndex], a, -Math.sign(getLineDist(points[maxIndex], a, b)));
        findHull(points, hull, points[maxIndex], b, -Math.sign(getLineDist(points[maxIndex], b, a)));

    }

}

let getHull = (points) => {

    let hull = new Set();

    let minX = 0,
        maxX = 0;

    for (let [index, point] of points.entries()) {

        if (point.x < points[minX].x) {
            minX = index;
        }

        if (point.x > points[maxX].x) {
            maxX = index;
        }

    }

    findHull(points, hull, points[minX], points[maxX], 1);
    findHull(points, hull, points[minX], points[maxX], -1);

    return hull;

}

let points = [
  {x: 0, y: 3},
  {x: 1, y: 1},
  {x: 2, y: 2},
  {x: 4, y: 4},
  {x: 0, y: 0},
  {x: 1, y: 2},
  {x: 3, y: 1},
  {x: 3, y: 3}
]

for(let p of getHull(points)) {
  console.log(p);
}
